service/matching/nexus_endpoint_client.go

```diff
@@ -360,3 +371,49 @@ func (m *nexusEndpointClient) resetCacheStateLocked() {
 	m.endpointsByID = make(map[string]*persistencespb.NexusEndpointEntry)
 	m.endpointsByName = make(map[string]*persistencespb.NexusEndpointEntry)
 }
+
+// notifyOwnershipChanged starts or stops a background routine which watches the Nexus endpoints table version for
+// changes. This is only expected to be called from matchingEngineImpl.notifyNexusEndpointsOwnershipChange()
+func (m *nexusEndpointClient) notifyOwnershipChanged(isOwner bool) {
+	m.refreshLock.Lock()
+	defer m.refreshLock.Unlock()
+
+	if isOwner && m.refreshHandle == nil {
+		// Just acquired ownership. Start refresh loop on table version to catch any updates from previous owner.
+		backgroundCtx := headers.SetCallerInfo(
+			context.Background(),
+			headers.SystemBackgroundCallerInfo,
+		)
+		m.refreshHandle = goro.NewHandle(backgroundCtx)
+		m.refreshHandle.Go(m.refreshTableVersion)
+	} else if !isOwner && m.refreshHandle != nil {
+		// Just lost ownership. Stop table version refresh loop.
+		m.refreshHandle.Cancel()
+		m.refreshHandle = nil
+	}
+}
+
+func (m *nexusEndpointClient) refreshTableVersion(ctx context.Context) error {
+	for ctx.Err() == nil {
+		m.checkTableVersion(ctx)
+		util.InterruptibleSleep(ctx, backoff.Jitter(m.endpointsRefreshInterval(), 0.2))
+	}
+	return ctx.Err()
+}
+
+func (m *nexusEndpointClient) checkTableVersion(ctx context.Context) {
+	// Acquire lock to make sure we are not in the middle of an update.
+	m.Lock()
+	defer m.Unlock()
```

Comment:
You should be able to lock after the persistence read request, that's slightly better.

Reply:
I think we need to do it before because an update may be in-progress and not reflected in our read from persistence, but when we acquire the lock the state has changed.

Reply:
So you could only refresh if the version we just read is `>`.

Not super critical, just could make the API a bit more responsive and not subject to DB availability in the common case.